app/services/db/monthly_sp_auth_count/unique_monthly_auth_counts_by_partner.rb

```diff
@@ -0,0 +1,143 @@
+
+        year_month_to_users_to_profile_age = Hash.new do |ym_h, ym_k|
+          ym_h[ym_k] = Hash.new { |user_h, user_k| }
+        end
+
+        queries.each do |query|
+          temp_copy = year_month_to_users_to_profile_age.deep_dup
+
+          with_retries(
+            max_tries: 3,
+            rescue: [
+              ActiveRecord::SerializationFailure,
+              PG::ConnectionBad,
+              PG::TRSerializationFailure,
+              PG::UnableToSend,
+            ],
+            handler: proc do
+              year_month_to_users_to_profile_age = temp_copy
+              ActiveRecord::Base.connection.reconnect!
+            end,
+          ) do
+            Reports::BaseReport.transaction_with_timeout do
+              ActiveRecord::Base.connection.execute(query).each do |row|
+                user_id = row['user_id']
+                year_month = row['year_month']
+                profile_age = row['profile_age']
+
+                year_month_to_users_to_profile_age[year_month][user_id] = profile_age
+              end
+            end
+          end
+        end
+
+        rows = []
+
+        prev_seen_users = Set.new
+        year_months = year_month_to_users_to_profile_age.keys.sort
+
+        year_months.each do |year_month|
+          users_to_profile_age = year_month_to_users_to_profile_age[year_month]
+
+          this_month_users = users_to_profile_age.keys.to_set
+          new_unique_users = this_month_users - prev_seen_users
+
+          profile_age_counts = new_unique_users.group_by do |user_id|
+            age = users_to_profile_age[user_id]
+            if age.nil? || age < 0
+              :unknown
+            elsif age.to_i > 4
+              :older
+            else
+              age
+            end
+          end
+
+          prev_seen_users |= this_month_users
+
+          rows << {
+            key: key,
+            year_month: year_month,
+            iaa_start_date: date_range.begin.to_s,
+            iaa_end_date: date_range.end.to_s,
+            unique_users: this_month_users.count,
+            new_unique_users: new_unique_users.count,
+            partner_ial2_new_unique_users_year1: (profile_age_counts[0].nil? ? 0 : profile_age_counts[0].count),
+            partner_ial2_new_unique_users_year2: (profile_age_counts[1].nil? ? 0 : profile_age_counts[1].count),
+            partner_ial2_new_unique_users_year3: (profile_age_counts[2].nil? ? 0 : profile_age_counts[2].count),
+            partner_ial2_new_unique_users_year4: (profile_age_counts[3].nil? ? 0 : profile_age_counts[3].count),
+            partner_ial2_new_unique_users_year5: (profile_age_counts[4].nil? ? 0 : profile_age_counts[4].count),
+            partner_ial2_new_unique_users_year_greater_than_5: profile_age_counts[:older].nil? ? 0 : profile_age_counts[:older].count,
+            partner_ial2_new_unique_users_unknown: profile_age_counts[:unknown].nil? ? 0 : profile_age_counts[:unknown].count,
```

Comment:
we could take advantage of setting `Hash#default` and simplify these checks:
```suggestion
          end.tap { |counts| counts.default = [] }

          prev_seen_users |= this_month_users

          rows << {
            key: key,
            year_month: year_month,
            iaa_start_date: date_range.begin.to_s,
            iaa_end_date: date_range.end.to_s,
            unique_users: this_month_users.count,
            new_unique_users: new_unique_users.count,
            partner_ial2_new_unique_users_year1: profile_age_counts[0].count,
            partner_ial2_new_unique_users_year2: profile_age_counts[1].count,
            partner_ial2_new_unique_users_year3: profile_age_counts[2].count,
            partner_ial2_new_unique_users_year4: profile_age_counts[3].count,
            partner_ial2_new_unique_users_year5: profile_age_counts[4].count,
            partner_ial2_new_unique_users_year_greater_than_5: profile_age_counts[:older].count,
            partner_ial2_new_unique_users_unknown: profile_age_counts[:unknown].count,
```